File: backend/src/main/java/com/reallink/pump/services/ProductSalesUnitChangeLogService.java
package com.reallink.pump.services;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.validation.annotation.Validated;

import com.reallink.pump.dto.request.CreateProductSalesUnitChangeLogRequest;
import com.reallink.pump.dto.response.ProductSalesUnitChangeLogResponse;
import com.reallink.pump.entities.Product;
import com.reallink.pump.entities.ProductSalesUnitChangeLog;
import com.reallink.pump.entities.ProductType;
import com.reallink.pump.exception.PumpBusinessException;
import com.reallink.pump.mapper.ProductSalesUnitChangeLogMapper;
import com.reallink.pump.repositories.ProductRepository;
import com.reallink.pump.repositories.ProductSalesUnitChangeLogRepository;

import jakarta.validation.Valid;
import jakarta.validation.constraints.NotNull;
import lombok.RequiredArgsConstructor;

@Service
@Validated
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class ProductSalesUnitChangeLogService {

    private final ProductSalesUnitChangeLogRepository changeLogRepository;
    private final ProductRepository productRepository;
    private final ProductSalesUnitChangeLogMapper mapper;

    /**
     * Get all change logs for a pump master
     */
    public List<ProductSalesUnitChangeLogResponse> getByPumpMasterId(@NotNull UUID pumpMasterId) {
        return changeLogRepository.findByPumpMaster_IdOrderByCreatedAtDesc(pumpMasterId).stream()
                .map(mapper::toResponse)
                .toList();
    }

    /**
     * Get change log by ID
     */
    public ProductSalesUnitChangeLogResponse getById(@NotNull UUID id) {
        ProductSalesUnitChangeLog changeLog = changeLogRepository.findById(id)
                .orElseThrow(() -> new PumpBusinessException("CHANGE_LOG_NOT_FOUND",
                "Change log with ID " + id + " not found"));
        return mapper.toResponse(changeLog);
    }

    /**
     * Get all change logs for a specific product
     */
    public List<ProductSalesUnitChangeLogResponse> getByProductId(@NotNull UUID productId) {
        return changeLogRepository.findByProduct_IdOrderByCreatedAtDesc(productId).stream()
                .map(mapper::toResponse)
                .toList();
    }

    /**
     * Get all change logs for a specific product type in a pump master
     */
    public List<ProductSalesUnitChangeLogResponse> getByPumpMasterIdAndProductType(
            @NotNull UUID pumpMasterId,
            @NotNull ProductType productType) {
        return changeLogRepository.findByPumpMaster_IdAndProductTypeOrderByCreatedAtDesc(pumpMasterId, productType).stream()
                .map(mapper::toResponse)
                .toList();
    }

    /**
     * Get change logs for a pump master within a date range
     */
    public List<ProductSalesUnitChangeLogResponse> getByPumpMasterIdAndDateRange(
            @NotNull UUID pumpMasterId,
            @NotNull LocalDateTime startDate,
            @NotNull LocalDateTime endDate) {
        return changeLogRepository.findByPumpMasterIdAndDateRange(pumpMasterId, startDate, endDate).stream()
                .map(mapper::toResponse)
                .toList();
    }

    /**
     * Get change logs for a specific product within a date range
     */
    public List<ProductSalesUnitChangeLogResponse> getByProductIdAndDateRange(
            @NotNull UUID productId,
            @NotNull LocalDateTime startDate,
            @NotNull LocalDateTime endDate) {
        return changeLogRepository.findByProductIdAndDateRange(productId, startDate, endDate).stream()
                .map(mapper::toResponse)
                .toList();
    }

    /**
     * Get all fuel product change logs within a date range for a pump master
     */
    public List<ProductSalesUnitChangeLogResponse> getFuelProductChangesByPumpMasterIdAndDateRange(
            @NotNull UUID pumpMasterId,
            @NotNull LocalDateTime startDate,
            @NotNull LocalDateTime endDate) {
        return changeLogRepository.findFuelProductChangesByPumpMasterIdAndDateRange(pumpMasterId, startDate, endDate).stream()
                .map(mapper::toResponse)
                .toList();
    }

    /**
     * Get the most recent change log for a product
     */
    public ProductSalesUnitChangeLogResponse getMostRecentChangeByProductId(@NotNull UUID productId) {
        ProductSalesUnitChangeLog changeLog = changeLogRepository.findMostRecentChangeByProductId(productId);
        return mapper.toResponse(changeLog);
    }

    /**
     * Create a new change log entry
     */
    @Transactional
    public ProductSalesUnitChangeLogResponse create(@Valid CreateProductSalesUnitChangeLogRequest request) {
        Product product = productRepository.findById(request.getProductId())
                .orElseThrow(() -> new PumpBusinessException("PRODUCT_NOT_FOUND",
                "Product with ID " + request.getProductId() + " not found"));

        ProductSalesUnitChangeLog changeLog = new ProductSalesUnitChangeLog(
                product,
                request.getOldSalesUnit(),
                request.getNewSalesUnit(),
                request.getOldStockQuantity(),
                request.getNewStockQuantity(),
                request.getOldSalesRate(),
                request.getNewSalesRate(),
                request.getChangeReason(),
                request.getChangedBy()
        );

        if (request.getRemarks() != null) {
            changeLog.setRemarks(request.getRemarks());
        }

        ProductSalesUnitChangeLog savedChangeLog = changeLogRepository.save(changeLog);
        return mapper.toResponse(savedChangeLog);
    }

    /**
     * Get count of change logs for a product
     */
    public long getCountByProductId(@NotNull UUID productId) {
        return changeLogRepository.countByProduct_Id(productId);
    }

    /**
     * Get count of change logs for a pump master
     */
    public long getCountByPumpMasterId(@NotNull UUID pumpMasterId) {
        return changeLogRepository.countByPumpMaster_Id(pumpMasterId);
    }

    /**
     * Check if a product has any change logs
     */
    public boolean existsByProductId(@NotNull UUID productId) {
        return changeLogRepository.existsByProduct_Id(productId);
    }
}

File: frontend/src/services/product-sales-unit-change-log-service.ts
import api from "./api";
import type {
  ProductSalesUnitChangeLog,
  CreateProductSalesUnitChangeLog,
} from "@/types/product-sales-unit-change-log";

const BASE_URL = "/api/product-sales-unit-change-logs";

export const productSalesUnitChangeLogService = {
  // Get all change logs for the current pump master
  getAll: async (): Promise<ProductSalesUnitChangeLog[]> => {
    const response = await api.get(BASE_URL);
    return response.data;
  },

  // Get change log by ID
  getById: async (id: string): Promise<ProductSalesUnitChangeLog> => {
    const response = await api.get(`${BASE_URL}/${id}`);
    return response.data;
  },

  // Get all change logs for a specific product
  getByProductId: async (
    productId: string
  ): Promise<ProductSalesUnitChangeLog[]> => {
    const response = await api.get(`${BASE_URL}/product/${productId}`);
    return response.data;
  },

  // Get all change logs for a specific product type
  getByProductType: async (
    productType: "FUEL" | "GENERAL"
  ): Promise<ProductSalesUnitChangeLog[]> => {
    const response = await api.get(`${BASE_URL}/product-type/${productType}`);
    return response.data;
  },

  // Get change logs within a date range
  getByDateRange: async (
    startDate: string,
    endDate: string
  ): Promise<ProductSalesUnitChangeLog[]> => {
    const response = await api.get(`${BASE_URL}/date-range`, {
      params: { startDate, endDate },
    });
    return response.data;
  },

  // Get change logs for a specific product within a date range
  getByProductIdAndDateRange: async (
    productId: string,
    startDate: string,
    endDate: string
  ): Promise<ProductSalesUnitChangeLog[]> => {
    const response = await api.get(
      `${BASE_URL}/product/${productId}/date-range`,
      {
        params: { startDate, endDate },
      }
    );
    return response.data;
  },

  // Get all fuel product change logs within a date range
  getFuelProductChangesByDateRange: async (
    startDate: string,
    endDate: string
  ): Promise<ProductSalesUnitChangeLog[]> => {
    const response = await api.get(`${BASE_URL}/fuel/date-range`, {
      params: { startDate, endDate },
    });
    return response.data;
  },

  // Get the most recent change log for a product
  getMostRecentChangeByProductId: async (
    productId: string
  ): Promise<ProductSalesUnitChangeLog> => {
    const response = await api.get(
      `${BASE_URL}/product/${productId}/most-recent`
    );
    return response.data;
  },

  // Create a new change log entry
  create: async (
    data: CreateProductSalesUnitChangeLog
  ): Promise<ProductSalesUnitChangeLog> => {
    const response = await api.post(BASE_URL, data);
    return response.data;
  },

  // Get count of change logs for a product
  getCountByProductId: async (productId: string): Promise<number> => {
    const response = await api.get(`${BASE_URL}/product/${productId}/count`);
    return response.data;
  },

  // Get count of change logs for the current pump master
  getCount: async (): Promise<number> => {
    const response = await api.get(`${BASE_URL}/count`);
    return response.data;
  },
};
